packages/test/test-utils/src/timeoutUtils.ts

```diff
@@ -4,15 +4,117 @@
+    private timer: NodeJS.Timeout | undefined;
+    private readonly deferred: Deferred<void>;
+    private rejected = false;
+
+    private static instance: TestTimeout;
+    public static initialize() {
+        TestTimeout.instance = new TestTimeout();
+    }
+    public static reset(runnable: Mocha.Runnable) {
+        TestTimeout.clear();
+        TestTimeout.instance.resetTimer(runnable);
+    }
+
+    public static clear() {
+        if (TestTimeout.instance.rejected) {
+            TestTimeout.instance = new TestTimeout();
+        } else {
+            TestTimeout.instance.clearTimer();
+        }
+    }
+
+    public static getInstance() {
+        return TestTimeout.instance;
+    }
+
+    public async getPromise() {
+        return this.deferred.promise;
+    }
+
+    public getTimeout() {
+        return this.timeout;
+    }
+
+    private constructor() {
+        this.deferred = new Deferred();
+        // Ignore rejection for timeout promise if no one is waiting for it.
+        this.deferred.promise.catch(() => { });
+    }
+
+    private resetTimer(runnable: Mocha.Runnable) {
+        assert(!this.timer, "clearTimer should have been called before reset");
+        assert(!this.deferred.isCompleted, "can't reset a completed TestTimeout");
+
+        // Check the test timeout setting
+        const timeout = runnable.timeout();
+        if (!(Number.isFinite(timeout) && timeout > 0)) { return; }
+
+        // subtract a buffer
+        this.timeout = Math.max(timeout - timeBuffer, 1);
+
+        // Set up timer to reject near the test timeout.
+        this.timer = setTimeout(() => {
+            this.deferred.reject(this);
+            this.rejected = true;
+        }, this.timeout);
+    }
+    private clearTimer() {
+        if (this.timer) {
+            clearTimeout(this.timer);
+            this.timer = undefined;
+        }
+    }
+}
+
+// only register if we are running with mocha-test-setup loaded
+if (globalThis.getMochaModule !== undefined) {
+    TestTimeout.initialize();
+
+    // patching resetTimeout and clearTimeout on the runnable object
+    // so we can track when test timeout are enforced
+    const mochaModule = globalThis.getMochaModule() as typeof Mocha;
+    const runnablePrototype = mochaModule.Runnable.prototype;
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    const oldResetTimeoutFunc = runnablePrototype.resetTimeout;
+    runnablePrototype.resetTimeout = function(this: Mocha.Runnable) {
+        oldResetTimeoutFunc.call(this);
+        TestTimeout.reset(this);
+    };
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    const oldClearTimeoutFunc = runnablePrototype.clearTimeout;
+    runnablePrototype.clearTimeout = function(this: Mocha.Runnable) {
+        TestTimeout.clear();
+        oldClearTimeoutFunc.call(this);
+    };
+}
+
 export interface TimeoutWithError {
+    // Timeout duration in milliseconds, if it is > 0 and not Infinity
```

Comment:
comments should be `/***` and use ts doc notation for docs and intellisense

Reply:
Done.